lib/Modules/HyperVCIHooks/HyperVCIHooks.psm1

```diff
@@ -0,0 +1,1538 @@
+            Test-FileIntegrity -File $OutFile -Algorithm $algorithm -ExpectedHash $hash
+        } else {
+            Throw "Hash algorithm $algorithm not recognized."
+        }
+    }
+}
+
+
+# Returns the full path of the package after it is downloaded using
+# the URL parameter (a checksum may optionally be specified). The
+# package is cached on the disk until the installation successfully finishes.
+# If the hook fails, on the second run this function will return the cached
+# package path if checksum is given and it matches.
+function Get-PackagePath {
+    Param(
+        [Parameter(Mandatory=$True)]
+        [string]$URL,
+        [string]$Checksum="",
+        [string]$HashingAlgorithm=""
+    )
+
+    $packagePath = Join-Path $env:TEMP $URL.Split('/')[-1]
+    if (Test-Path $packagePath) {
+        if ($Checksum -and $HashingAlgorithm) {
+            if (Test-FileIntegrity -File $packagePath -Algorithm $HashingAlgorithm -ExpectedHash $Checksum) {
+                return $packagePath
+            }
+        }
+        Remove-Item -Recurse -Force -Path $packagePath
+    }
+
+    if ($Checksum -and $HashingAlgorithm) {
+        Start-DownloadFile -Uri "$URL#$HashingAlgorithm=$Checksum" -OutFile $packagePath
+    } else {
+        Start-DownloadFile -Uri $URL -OutFile $packagePath -SkipIntegrityCheck
+    }
+
+    return $packagePath
+}
+
+
+# Installs a package after it is downloaded from the Internet and checked for
+# integrity with SHA1 checksum. Accepts as parameters: an URL, an optional
+# 'Checksum' with its 'HashingAlgorithm' and 'ArgumentList' which can be passed
+# if the installer requires unattended installation.
+# Supported packages formats are: '.exe' and '.msi'
+function Install-Package {
+    Param(
+        [string]$URL,
+        [string]$Checksum="",
+        [string]$HashingAlgorithm="",
+        [array]$ArgumentList
+    )
+
+    Write-JujuLog "Installing package '$URL'"
+
+    $packageFormat = $URL.Split('.')[-1]
+    $acceptedFormats = @('msi', 'exe')
+    if ($packageFormat -notin $acceptedFormats) {
+        Throw ("Cannot install the package found at this URL: $URL " +
+               "Unsupported installer format.")
+    }
+
+    $installerPath = Get-PackagePath $URL $Checksum $HashingAlgorithm
+    $stat = Start-Process -FilePath $installerPath -ArgumentList $ArgumentList `
+                          -PassThru -Wait
+    if ($stat.ExitCode -ne 0) {
+        throw "Package '$URL' failed to install."
+    }
+    Remove-Item $installerPath
+
+    Write-JujuLog "Finished installing package."
+}
+
+
+function Start-GitClonePull {
+    Param(
+        [string]$Path,
+        [string]$URL,
+        [string]$Branch="master"
+    )
+
+    if (!(Test-Path -Path $Path)) {
+        Start-ExecuteWithRetry {
+            Start-ExternalCommand -ScriptBlock { git clone $URL $Path } `
+                                  -ErrorMessage "Git clone failed"
+        }
+        pushd $Path
```

Comment:
git has a nice ```-C``` option that says:

```
-C <path>
           Run as if git was started in <path> instead of the current working
           directory. When multiple -C options are given, each subsequent
           non-absolute -C <path> is interpreted relative to the preceding -C
           <path>.

           This option affects options that expect path name like --git-dir
           and --work-tree in that their interpretations of the path names
           would be made relative to the working directory caused by the -C
           option. For example the following invocations are equivalent:

               git --git-dir=a.git --work-tree=b -C c status
               git --git-dir=c/a.git --work-tree=c/b status
```

So you could just run one ```Start-ExternalCommand``` with the following script block:

```
Start-ExecuteWithRetry {
        Start-ExternalCommand -ScriptBlock {
                                  git clone $URL $Path
                                  git -C $Path checkout $Branch
                          } `
                                  -ErrorMessage "Git clone failed"
        }
```

And you can remove the pushd